Added Header Authentication - Client Part

joewashear007 opened the pull request:
Client Side modifications for https://github.com/actualbudget/actual-server/pull/312 for https://github.com/actualbudget/actual/issues/1092

This change added the check for the "loginMethod" from the `/account/use-bootstrap` endpoint. It will then adjust the login page to automatically call the login endpoint if the loginMethod == "header". 

I was having trouble getting this built and tested, so I would appreciate pointers or testing tips. 

Thanks!

youngcw commented:
If the server gets set with the 'header' env var does that block using the password or just allow headers?

youngcw commented:
Are you testing this with Authentik?  Could you write up your steps to get that setup?  Im missing something in how that should work

joewashear007 commented:
I run authentik in docker with is pretty straight forward, but my setup also runs with nginx-proxy as my main reverse proxy. So my setup assumes that, other reverse proxy providers could be used. I also had to setup the main urls and such in various places.

1. I setup a new  Proxy Provider and set it to "Forward Auth", since I have nginx proxy manager [instruction](https://goauthentik.io/docs/providers/proxy/server_nginx). I updated my nginx.conf with the auth and proxy_pass headers as in the documentation. 
2. I created a new Application for actual budget
3. I updated the Embedded Output to select the proxy provider that I created. 
4. I created a new "finance" group and added  following to the attributes block
```
additionalHeaders:
  X-ACTUAL-PASSWORD: <redacted>
```
5. I assigned the "finance" group to actual budget application

Once this is in place, any call to my actual instance get checked for valid Authentik logins (through the magic of nginx). Authentik should then automagically pass on  X-ACTUAL-PASSWORD header to my actual instance.

I hope that is good enough descriptions, I am happy to give my details on my setup, but not in the public github thread. I also haven't been able to test this end to end fully yet since I wasn't sure how to build a docker image that had my change from both the server and the client repo.

joewashear007 commented:
I am still trying to find time to work on for https://github.com/actualbudget/actual-server/pull/312.

twk3 commented on packages/desktop-client/src/components/manager/subscribe/Login.tsx:
```diff
@@ -86,27 +119,52 @@ export function Login() {
+            autoFocus={true}
+            placeholder="Password"
+            type="password"
+            onChange={e => setPassword(e.target.value)}
```
```suggestion
            onChange={(e: ChangeEvent<HTMLInputElement>) =>
              setPassword(e.target.value)
            }
```

To match the original

twk3 commented on upcoming-release-notes/2362.md:
```diff
@@ -0,0 +1,6 @@
+authors: [joewashear007]
+---
+
+Add option to authenticate with HTTP header from Auth Proxy. Set new env `ACTUAL_LOGIN_METHOD` = 'header' and make sure your proxy set the `X-ACTUAL-PASSWORD` to your actual password. This front end will attempt to auto login into actual if the header is accepted by the server. If the header is not valid or incorrect, the user is notified and given the option to use their password to login. Pleas ...
```
Release note should be just the first sentence here.

joewashear007 replied:
Done

twk3 commented on packages/desktop-client/src/components/manager/subscribe/Login.tsx:
```diff
@@ -1,13 +1,16 @@
 // @ts-strict-ignore
-import React, { type ChangeEvent, useState } from 'react';
+import { useState, useEffect } from 'react';
```
```suggestion
import React, { type ChangeEvent, useState, useEffect } from 'react';
```

Missing some changes from the default branch

twk3 commented on packages/desktop-client/src/components/manager/subscribe/Login.tsx:
```diff
@@ -1,13 +1,16 @@
+import { AnimatedLoading } from '../../../icons/AnimatedLoading';
 import { theme } from '../../../style';
 import { Button, ButtonWithLoading } from '../../common/Button';
+import { ButtonLink } from '../../common/ButtonLink';
```
```suggestion
import { ButtonLink } from '../../common/Link';
```

These were consolidated into one file

joewashear007 commented on packages/desktop-client/src/components/manager/subscribe/common.tsx:
```diff
@@ -50,7 +52,7 @@ export function useBootstrapped() {
 
         if (result.bootstrapped) {
-          ensure('/login');
+          ensure(`/login/${result.loginMethod}`);
```
Sounds perfectly resonable

joewashear007 commented on packages/loot-core/src/server/main.ts:
```diff
@@ -1492,7 +1500,7 @@ handlers['subscribe-sign-in'] = async function ({ password }) {
   }
 
-  return { error: 'invalid-password' };
+  return { error: res.error || 'invalid-password' };
```
```suggestion
  return { error: res.reason|| 'invalid-password' };
```
@twk3 - changing this to reason since that is what this commit now is returning: https://github.com/actualbudget/actual-server/commit/1bbba663b2c20b722327d885d0cbf80dba0ea44d#diff-024ac510e5825a4edf4be0b860f4abba35556d3ca0b5aad96c0499ed126f9452R44

twk3 commented:
@joewashear007 both PRs look ready for one final review (other than the merge conflict). And then I will merge. But now that we are in mergefreeze for the release, I am going to hold off on the final review until closer to when the freeze is lifted, so the review and a merge sanity check can happen at the same time. 

Likely the end of next week, or over that weekend.

mioiox commented:
Thanks for bringing up this feature. I personally can confirm it's an important one since I have a Kemp Virtual Load Master (VLM) in front of Actual with LDAP preauthentication enabed on the VLM  - and I was struggling to make it work. So this header passing the "authenticated" flag to Actual server is a super cool feature. I am looking forward to it. 
@twk3, do you have an idea if it will make it for the next release and when this release will take place?

joewashear007 commented:
Thank you @twk3 - I will be pretty excited to get this merged

twk3 reviewed (approved):
Ran a final round of testing today. Looks good.

There is some weirdness around signout when using the header (as we just sign back in). But the fact that we have the change server menu option, allows you to instead use that to 'sign out'. 

Thanks @joewashear007 for your long work on this contribution

Shazib commented:
Sorry to necro @joewashear007, Can you confirm if your proxy setup allows you to use the desktop app and web app without any sever connection issues?

I've using Authelia + Clourflare tunnels and have trouble getting the desktop app to work as the calls to the API actually return a 302 redirect which at the moment the `fetch` calls don't handle well.

joewashear007 commented:
@Shazib - so the PWA/service worker network caching for offline usage is a separate issue (which is quite annoying). 

This PR works with passing the password via hreader from the proxy. A fix for the offline usage with an auth proxy would need to happen with the service workers. It needs a URL endpoint that can detect if the proxy is expired that is not cached. A work around that I have been using on desktop browsers is to unregister the service worker and then reload.

I would like to make a PR to fix this, but I haven't gotten the time yet

Shazib commented:
Thanks for the info @joewashear007!

I will give adding the header a go. 

For now, to solve the other issue, I've added a bypass rule using cloudflare's 'warp' client. So if my specific PC tries to access my actual instance it bypasses the auth layer entirely.

I guess in a perfect would we would add the OAuth2 capabilities directly to Actual. in the browser the auth redirect works ok, but the desktop app needs additional handling.

max-tet commented:
Is there a way to completely remove the need to set a password?
As far as I understand, this feature allows you to pass the password in a header value but when accessing the server for the first time, a password must still be set, right?

For a hosting setup using a reverse proxy and auth provider, this should also be unneccessary. In fact, IMO there should be no password at all.

Shazib commented:
> Is there a way to completely remove the need to set a password? As far as I understand, this feature allows you to pass the password in a header value but when accessing the server for the first time, a password must still be set, right?
> 
> For a hosting setup using a reverse proxy and auth provider, this should also be unneccessary. In fact, IMO there should be no password at all.

This sort of circles back to my original query, and would require Actual to natively support OIDC so it can pass auth responsibilities off to your auth provider. 

This is not something that is currently supported so someone would need to implement it and submit a PR, it will likely not be a priority for the core team.
